core/common/src/main/java/alluxio/underfs/ObjectLowLevelOutputStream.java

```diff
@@ -0,0 +1,410 @@
+    if (mFile == null) {
+      return;
+    }
+    mLocalOutputStream.close();
+    int partNumber = mPartNumber.getAndIncrement();
+    File newFileToUpload = new File(mFile.getPath());
+    mFile = null;
+    mLocalOutputStream = null;
+    uploadPart(newFileToUpload, partNumber, false);
+  }
+
+  protected void uploadPart(File file, int partNumber, boolean lastPart) {
+    final String md5;
+    if (mHash != null) {
+      md5 = Base64.encodeBase64String(mHash.digest());
+    } else {
+      md5 = null;
+    }
+    Callable<?> callable = () -> {
+      IOException lastException = null;
+      try {
+        do {
+          try {
+            uploadPartInternal(file, partNumber, lastPart, md5);
+            return null;
+          } catch (IOException e) {
+            lastException = e;
+          }
+        } while (mRetryPolicy.attempt());
+      } finally {
+        // Delete the uploaded or failed to upload file
+        if (!file.delete()) {
+          LOG.error("Failed to delete temporary file @ {}", file.getPath());
+        }
+      }
+      throw new IOException("Fail to upload part " + partNumber + " to " + mKey, lastException);
+    };
+    ListenableFuture<?> futureTag = mExecutor.submit(callable);
+    mFutures.add(futureTag);
+    LOG.debug(
+        "Submit upload part request. key={}, partNum={}, file={}, fileSize={}, lastPart={}.",
+        mKey, partNumber, file.getPath(), file.length(), lastPart);
+  }
+
+  protected void abortMultiPartUpload() {
+    IOException lastException;
+    do {
+      try {
+        abortMultiPartUploadInternal();
+        LOG.warn("Aborted multipart upload for key {} and  to bucket {}",
+            mKey, mBucketName);
+        return;
+      } catch (IOException e) {
+        lastException = e;
+      }
+    } while (mRetryPolicy.attempt());
+    // This point is only reached if the operation failed more
+    // than the allowed retry count
+    LOG.warn("Unable to abort multipart upload for key '{}' to bucket {}. "
+            + "You may need to enable the periodical cleanup by setting property {}"
+            + "to be true.", mKey, mBucketName,
+        PropertyKey.UNDERFS_CLEANUP_ENABLED.getName(),
+        lastException);
+  }
+
+  protected void waitForAllPartsUpload() throws IOException {
+    try {
+      for (ListenableFuture<?> future : mFutures) {
+        future.get();
+      }
+    } catch (ExecutionException e) {
+      // No recover ways so that we need to cancel all the upload tasks
+      // and abort the multipart upload
+      Futures.allAsList(mFutures).cancel(true);
+      abortMultiPartUpload();
+      throw new IOException(
+          "Part upload failed in multipart upload with to " + mKey, e);
+    } catch (InterruptedException e) {
+      LOG.warn("Interrupted object upload.", e);
+      Futures.allAsList(mFutures).cancel(true);
+      abortMultiPartUpload();
+      Thread.currentThread().interrupt();
+    }
+    mFutures = new ArrayList<>();
```

Comment:
Would it be better to replace it with `mFutures.clear()` to avoid using more memory?

Reply:
fix.